File: server/src/main/java/com/smartexpense/api/application/usecase/user/ProfileUseCaseImpl.java
package com.smartexpense.api.application.usecase.user;

import com.smartexpense.api.application.dto.request.UpdateProfileRequestDto;
import com.smartexpense.api.application.dto.response.UserProfileResponseDto;
import com.smartexpense.api.application.port.in.ProfileUseCase;
import com.smartexpense.api.domain.model.User;
import com.smartexpense.api.domain.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
public class ProfileUseCaseImpl implements ProfileUseCase {

    private final UserRepository userRepository;

    @Override
    public UserProfileResponseDto getProfile(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return mapToDto(user);
    }

    @Override
    public UserProfileResponseDto updateProfile(String email, UpdateProfileRequestDto request) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (request.getDisplayName() != null) {
            user.setDisplayName(request.getDisplayName());
        }
        if (request.getAvatarUrl() != null) {
            user.setAvatarUrl(request.getAvatarUrl());
        }
        if (request.getMonthlyBudget() != null) {
            user.setMonthlyBudget(request.getMonthlyBudget());
        }
        if (request.getThemeColor() != null) {
            user.setThemeColor(request.getThemeColor());
        }
        
        user.setUpdatedAt(LocalDateTime.now());
        User savedUser = userRepository.save(user);

        return mapToDto(savedUser);
    }

    private UserProfileResponseDto mapToDto(User user) {
        return UserProfileResponseDto.builder()
                .id(user.getId())
                .email(user.getEmail())
                .displayName(user.getDisplayName())
                .avatarUrl(user.getAvatarUrl())
                .monthlyBudget(user.getMonthlyBudget())
                .streakDays(user.getStreakDays())
                .themeColor(user.getThemeColor())
                .build();
    }
}

File: server/bin/main/com/smartexpense/api/application/usecase/user/ProfileUseCaseImpl.kt
package com.smartexpense.api.application.usecase.user

import com.smartexpense.api.application.dto.request.UpdateProfileRequestDto
import com.smartexpense.api.application.dto.response.UserProfileResponseDto
import com.smartexpense.api.application.port.`in`.ProfileUseCase
import com.smartexpense.api.domain.model.User
import com.smartexpense.api.domain.repository.UserRepository
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class ProfileUseCaseImpl(
    private val userRepository: UserRepository
) : ProfileUseCase {

    override fun getProfile(email: String): UserProfileResponseDto {
        val user = userRepository.findByEmail(email)
            ?: throw RuntimeException("User not found")
        return mapToDto(user)
    }

    override fun updateProfile(email: String, request: UpdateProfileRequestDto): UserProfileResponseDto {
        val user = userRepository.findByEmail(email)
            ?: throw RuntimeException("User not found")

        val updatedUser = user.copy(
            displayName = request.displayName ?: user.displayName,
            avatarUrl = request.avatarUrl ?: user.avatarUrl,
            monthlyBudget = request.monthlyBudget ?: user.monthlyBudget,
            themeColor = request.themeColor ?: user.themeColor,
            updatedAt = LocalDateTime.now()
        )

        return mapToDto(userRepository.save(updatedUser))
    }

    private fun mapToDto(user: User) = UserProfileResponseDto(
        id = user.id!!,
        email = user.email,
        displayName = user.displayName,
        avatarUrl = user.avatarUrl,
        monthlyBudget = user.monthlyBudget,
        streakDays = user.streakDays ?: 0,
        themeColor = user.themeColor ?: "#FFE600"
    )
}
